SYCL/Basic/device_implicitly_copyable.cpp

```diff
@@ -0,0 +1,509 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+// RUN: %ACC_RUN_PLACEHOLDER %t.out
+
+//==--device_implicitly_copyable.cpp - SYCL implicit device copyable test --==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#include <cassert>
+#include <iostream>
+
+#include <sycl/sycl.hpp>
+
+struct ACopyable {
+  int i;
+  ACopyable() = default;
+  ACopyable(int _i) : i(_i) {}
+  ACopyable(const ACopyable &x) : i(x.i) {}
+};
+
+template <> struct sycl::is_device_copyable<ACopyable> : std::true_type {};
+
+int main() {
+  constexpr size_t arr_size = 5;
+  constexpr int ref_val = 14;
+  sycl::queue q;
+  {
+    std::pair<int, float> pair_arr[arr_size];
+    std::pair<int, float> pair{ref_val, ref_val};
+    std::pair<int, float> result_pair_arr[arr_size];
+    std::pair<int, float> result_pair;
+
+    for (auto i = 0; i < arr_size; i++) {
+      pair_arr[i].first = i;
+      pair_arr[i].second = i;
+    }
+
+    {
+      sycl::buffer<std::pair<int, float>, 1> buf_pair_arr{
+          result_pair_arr, sycl::range<1>(arr_size)};
+      sycl::buffer<std::pair<int, float>, 1> buf_pair{&result_pair,
+                                                      sycl::range<1>(1)};
+
+      q.submit([&](sycl::handler &cgh) {
+         auto acc_pair_arr =
+             sycl::accessor{buf_pair_arr, cgh, sycl::read_write};
+         auto acc_pair = sycl::accessor{buf_pair, cgh, sycl::read_write};
+         cgh.single_task([=]() {
+           for (auto i = 0; i < arr_size; i++) {
+             acc_pair_arr[i] = pair_arr[i];
+           }
+           acc_pair[0] = pair;
+         });
+       }).wait_and_throw();
+    }
```

Comment:
This looks like it is the same for most of the test cases, except for the type used. Could it be moved to a common function? I'm thinking something like:

```suggestion
template <typename DataT, size_t ArrSize>
void CaptureAndCopy(const DataT *data_arr, const DataT &data_scalar, DataT *result_arr, DataT *result_scalar) {
    sycl::buffer<DataT, 1> buf_arr{result_arr, sycl::range<1>(ArrSize)};
    sycl::buffer<DataT, 1> buf_scalar{&result_scalar, sycl::range<1>(1)};

    q.submit([&](sycl::handler &cgh) {
       auto acc_arr = sycl::accessor{buf_arr, cgh, sycl::read_write};
       auto acc_scalar = sycl::accessor{buf_scalar, cgh, sycl::read_write};
       cgh.single_task([=]() {
         for (auto i = 0; i < ArrSize; i++) {
           acc_arr[i] = data_arr[i];
         }
         acc_scalar[0] = data_scalar;
       });
     });
  }
```